### ISSUE TITLE
Problems with GraphSage

### ISSUE THREAD

**Issue by user4443** [danielegrattarola/spektral#20]
Hi, thanks for the great package!, I'm trying to train embedding of graph vertices through link prediction using GraphSage conv.
Usually training goes without problems, but once every few hundred batches this error appears:

> 0 1614 train_epoch_loss 0.12077645 train_epoch_acc 0.83266443
> 0 1615 train_epoch_loss 0.12074445 train_epoch_acc 0.8327161
> 0 1616 train_epoch_loss 0.120728664 train_epoch_acc 0.832744
> 0 1617 train_epoch_loss 0.1207294 train_epoch_acc 0.8327376
> 0 1618 train_epoch_loss 0.12071675 train_epoch_acc 0.83276695
> 0 1619 train_epoch_loss 0.120711125 train_epoch_acc 0.8327747

> 0 1620 train_epoch_loss 0.1206873 train_epoch_acc 0.8328143
> ConcatOp : Dimensions of inputs should match: shape[0] = [94,200] vs. shape[1] = [93,200]
> 	 [[node graph_sage_conv_1/concat (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:1956) ]]
> 	 [[node graph_sage_conv_3/GatherV2 (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/spektral/layers/convolutional.py:378) ]]
> 
> Caused by op 'graph_sage_conv_1/concat', defined at:
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel_launcher.py", line 16, in <module>
>     app.launch_new_instance()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/traitlets/config/application.py", line 658, in launch_instance
>     app.start()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelapp.py", line 505, in start
>     self.io_loop.start()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/platform/asyncio.py", line 148, in start
>     self.asyncio_loop.run_forever()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/asyncio/base_events.py", line 539, in run_forever
>     self._run_once()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/asyncio/base_events.py", line 1775, in _run_once
>     handle._run()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/asyncio/events.py", line 88, in _run
>     self._context.run(self._callback, *self._args)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/ioloop.py", line 690, in <lambda>
>     lambda f: self._run_callback(functools.partial(callback, future))
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
>     ret = callback()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 781, in inner
>     self.run()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 742, in run
>     yielded = self.gen.send(value)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelbase.py", line 365, in process_one
>     yield gen.maybe_future(dispatch(*args))
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelbase.py", line 272, in dispatch_shell
>     yield gen.maybe_future(handler(stream, idents, msg))
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelbase.py", line 542, in execute_request
>     user_expressions, allow_stdin,
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/ipkernel.py", line 294, in do_execute
>     res = shell.run_cell(code, store_history=store_history, silent=silent)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/zmqshell.py", line 536, in run_cell
>     return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 2848, in run_cell
>     raw_cell, store_history, silent, shell_futures)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 2874, in _run_cell
>     return runner(coro)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/async_helpers.py", line 67, in _pseudo_sync_runner
>     coro.send(None)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3049, in run_cell_async
>     interactivity=interactivity, compiler=compiler, result=result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3214, in run_ast_nodes
>     if (yield from self.run_code(code, result)):
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3296, in run_code
>     exec(code_obj, self.user_global_ns, self.user_ns)
>   File "<ipython-input-34-85c905d3b7fa>", line 30, in <module>
>     x = GraphSageConv(bb_embed_size)([x, A_in])
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/engine/base_layer.py", line 457, in __call__
>     output = self.call(inputs, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/spektral/layers/convolutional.py", line 380, in call
>     output = K.concatenate([features, features_neigh])
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py", line 1956, in concatenate
>     return tf.concat([to_dense(x) for x in tensors], axis)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py", line 180, in wrapper
>     return target(*args, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/ops/array_ops.py", line 1256, in concat
>     return gen_array_ops.concat_v2(values=values, axis=axis, name=name)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1149, in concat_v2
>     "ConcatV2", values=values, axis=axis, name=name)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
>     op_def=op_def)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
>     return func(*args, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
>     op_def=op_def)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
>     self._traceback = tf_stack.extract_stack()
> 
> InvalidArgumentError (see above for traceback): ConcatOp : Dimensions of inputs should match: shape[0] = [94,200] vs. shape[1] = [93,200]
> 	 [[node graph_sage_conv_1/concat (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:1956) ]]
> 	 [[node graph_sage_conv_3/GatherV2 (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/spektral/layers/convolutional.py:378) ]]
> 
> 0 1622 train_epoch_loss 0.12065845 train_epoch_acc 0.83286613
> 0 1623 train_epoch_loss 0.120651215 train_epoch_acc 0.83288383
> 0 1624 train_epoch_loss 0.12064571 train_epoch_acc 0.8328906
> 0 1625 train_epoch_loss 0.12062876 train_epoch_acc 0.83291864
> ConcatOp : Dimensions of inputs should match: shape[0] = [159,200] vs. shape[1] = [158,200]
> 	 [[node graph_sage_conv_1/concat (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:1956) ]]
> 	 [[node graph_sage_conv_3/GatherV2 (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/spektral/layers/convolutional.py:378) ]]
> 
> Caused by op 'graph_sage_conv_1/concat', defined at:
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel_launcher.py", line 16, in <module>
>     app.launch_new_instance()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/traitlets/config/application.py", line 658, in launch_instance
>     app.start()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelapp.py", line 505, in start
>     self.io_loop.start()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/platform/asyncio.py", line 148, in start
>     self.asyncio_loop.run_forever()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/asyncio/base_events.py", line 539, in run_forever
>     self._run_once()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/asyncio/base_events.py", line 1775, in _run_once
>     handle._run()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/asyncio/events.py", line 88, in _run
>     self._context.run(self._callback, *self._args)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/ioloop.py", line 690, in <lambda>
>     lambda f: self._run_callback(functools.partial(callback, future))
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
>     ret = callback()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 781, in inner
>     self.run()
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 742, in run
>     yielded = self.gen.send(value)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelbase.py", line 365, in process_one
>     yield gen.maybe_future(dispatch(*args))
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelbase.py", line 272, in dispatch_shell
>     yield gen.maybe_future(handler(stream, idents, msg))
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/kernelbase.py", line 542, in execute_request
>     user_expressions, allow_stdin,
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tornado/gen.py", line 209, in wrapper
>     yielded = next(result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/ipkernel.py", line 294, in do_execute
>     res = shell.run_cell(code, store_history=store_history, silent=silent)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/ipykernel/zmqshell.py", line 536, in run_cell
>     return super(ZMQInteractiveShell, self).run_cell(*args, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 2848, in run_cell
>     raw_cell, store_history, silent, shell_futures)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 2874, in _run_cell
>     return runner(coro)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/async_helpers.py", line 67, in _pseudo_sync_runner
>     coro.send(None)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3049, in run_cell_async
>     interactivity=interactivity, compiler=compiler, result=result)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3214, in run_ast_nodes
>     if (yield from self.run_code(code, result)):
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/IPython/core/interactiveshell.py", line 3296, in run_code
>     exec(code_obj, self.user_global_ns, self.user_ns)
>   File "<ipython-input-34-85c905d3b7fa>", line 30, in <module>
>     x = GraphSageConv(bb_embed_size)([x, A_in])
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/engine/base_layer.py", line 457, in __call__
>     output = self.call(inputs, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/spektral/layers/convolutional.py", line 380, in call
>     output = K.concatenate([features, features_neigh])
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py", line 1956, in concatenate
>     return tf.concat([to_dense(x) for x in tensors], axis)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/util/dispatch.py", line 180, in wrapper
>     return target(*args, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/ops/array_ops.py", line 1256, in concat
>     return gen_array_ops.concat_v2(values=values, axis=axis, name=name)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/ops/gen_array_ops.py", line 1149, in concat_v2
>     "ConcatV2", values=values, axis=axis, name=name)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
>     op_def=op_def)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
>     return func(*args, **kwargs)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 3300, in create_op
>     op_def=op_def)
>   File "/home/user/miniconda3/envs/safe/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
>     self._traceback = tf_stack.extract_stack()
> 
> InvalidArgumentError (see above for traceback): ConcatOp : Dimensions of inputs should match: shape[0] = [159,200] vs. shape[1] = [158,200]
> 	 [[node graph_sage_conv_1/concat (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:1956) ]]
> 	 [[node graph_sage_conv_3/GatherV2 (defined at /home/user/miniconda3/envs/safe/lib/python3.7/site-packages/spektral/layers/convolutional.py:378) ]]
> 
> 0 1627 train_epoch_loss 0.12060276 train_epoch_acc 0.83297145
> 0 1628 train_epoch_loss 0.120608345 train_epoch_acc 0.83296674
> 0 1629 train_epoch_loss 0.12059478 train_epoch_acc 0.832987
> 0 1630 train_epoch_loss 0.1205696 train_epoch_acc 0.83303547
> 0 1631 train_epoch_loss 0.120563745 train_epoch_acc 0.83303756
> 0 1632 train_epoch_loss 0.12054803 train_epoch_acc 0.833069
> 0 1633 train_epoch_loss 0.12054562 train_epoch_acc 0.8330679`

Loss decreases, acc increases. It seems the network is learning.
I control all the dimensions of my data through assert.
Tell me what could be the problem?

**Comment by danielegrattarola** [comment]
Hi, 

I think this can happen if you have a node in your graph that is completely disconnected (its row and column in the adjacency matrix will be all zeros). 
Can you check if this is the case?

Cheers

**Comment by user4443** [comment]
Super! I'll check it now!

**Comment by user4443** [comment]
No. Isolated vertices do not throw an exception.

**Comment by danielegrattarola** [comment]
Well, the issue comes from these two statements in the GraphSage layer:

```python
features_neigh = self.aggregate_op(
    tf.gather(features, fltr.indices[:, -1]), fltr.indices[:, -2]
)
output = K.concatenate([features, features_neigh])
```

The crash is due to the fact that the input features given to the layer (`features`) and the aggregated features of each node's neighbour (`features_neigh`) don't have the same number of rows. 
This is due to an empty row or empty column in the adjacency matrix. 

If you have your adjacency matrix `A` in COO format (with `N` nodes), then the following should be true: 

- `all( np.unique(A.row) == np.arange(N) )`
- `all( np.unique(A.col) == np.arange(N) )`

If you still can't fix this, please post a minimal code sample to reproduce the issue.

**Comment by user4443** [comment]
Thank you very much for your reply! It looks like it is! I just began to skip examples where several connected components and errors were no more.
